lib/rubocop/cop/style/parallel_assignment.rb

```diff
@@ -69,10 +77,53 @@ def autocorrect(node)
 
         private
 
-        def swapping_variables?(left_elements, right_elements)
-          left_elements.any? do |le|
-            right_elements.any? do |re|
-              re.loc.expression.is?(le.loc.expression.source)
+        def find_valid_order(left_elements, right_elements)
+          # arrange left_elements in an order such that no corresponding right
+          # element refers to a left element earlier in the sequence
+          # this can be done using an algorithm called a "topological sort"
+          # fortunately for us, Ruby's stdlib contains an implementation
+          assignments = left_elements.zip(right_elements)
+
+          begin
+            AssignmentSorter.new(assignments).tsort
+          rescue TSort::Cyclic
+            nil
+          end
+        end
+
+        # Helper class necessitated by silly design of TSort prior to Ruby 2.1
+        # Newer versions have a better API, but that doesn't help us, does it?
+        class AssignmentSorter
+          include TSort
+
+          def initialize(assignments)
+            @assignments = assignments
+          end
+
+          def tsort_each_node
+            @assignments.each { |a| yield a }
+          end
+
+          def tsort_each_child(assignment)
+            # yield all the assignments which must come after `assignment`
+            # (due to dependencies on the previous value of the assigned var)
+            my_lhs, _my_rhs = *assignment
+
+            find_symbol = lambda do |node, symbol|
+              node.children.any? { |obj| obj.equal?(symbol) } ||
+              node.each_child_node.any? { |child| find_symbol[child, symbol] }
+            end
+
+            @assignments.each do |other|
+              next if other.equal?(assignment)
```

Comment:
@jonas054, it turns out that `TSort` has no problem with nodes which are their own children. I thought that would cause the topo sort to fail -- it doesn't. So I have removed that line. (But I did add one extra test case.)